### PR TITLE
Adding support for TLS and client certificated on MQTT broker

### PR DESCRIPTION
**Pull request by jamesonuk**
As requested, this is a pull request to allow MQTTS and client certificates to be used.
I have tested this with my live setup and also with a standard MQTT server using username / password.
I do however only have an alarm so I have not been able to test any camera details but as you can see this is simply a few extra variables that get passed through to the MQTT connection

### CODE REVIEW

**Comment by tsightler** [comment]
Hi @jamesonuk.  Thanks for the PR, I think this probably needs some additional changes.  I'm not concerned about cameras breaking because these changes are limited to MQTT connection so it will either work or not, however, standard installation, as well as the addon, uses a config file and it's not clear to me that this PR addresses that use case.  Even if it works, at a minimum, the docs for standard installation will require an update as well and then those changes need to be duplicated into the HA addon as well.

I won't have much time to look over this in the coming weeks, so it may be a little while.

**Comment by tsightler** [comment]
I'm going to close this PR because I have decided to stop maintaining this project for now.  I'm not sure if I'm going to come back to it or not, but I will at least be taking a break for a few months.  I do appreciate your contribution and if I do come back to this project I will reconsider integrating your changes.

**Comment by jonootto** [comment]
Is this something you'd consider reopening?

File: docs/DOCKER.md [hunk 1]
```diff
@@ -40,6 +40,10 @@ services:
       - MQTTPORT=1883                  # TCP port for MQTT Broker
       - MQTTUSER=mqtt_user             # CHANGE ME -- Username for MQTT Broker (remove for anonymous)
       - MQTTPASSWORD=mqtt_password     # CHANGE ME -- Password for MQTT Broker (remove for anonymous)
+      - MQTTPROTOCOL=                  # Set to mqtts if using a secure connection
+      - MQTTCLIENTKEY=                 # Set to path of the key file if using client certificate authentication
+      - MQTTCLIENTCERT-                # Set to path of the cert file if using client certificate authentication
+      - MQTTCA=                        # Set to path of CA pem file if using self signed certificate
```
Comment (jonootto) [inline docs/DOCKER.md:46]: Would be great to add option for self signed using rejectUnauthorized: false https://github.com/mqttjs/MQTT.js

**Comment by tsightler** [comment]
In the future, maybe, but not any time soon.  This PR just doesn't cover enough of the use cases and I don't feel like dealing with all of the issues at this time for a small percentage of users.

**Comment by tsightler** [comment]
Just to clarify, I'm not at all opposed to the idea of implementing this, I just don't want to implement something that doesn't address all of the possible install types and their various configuration methods because that doesn't really get this off the to-do list.

The reality is, I probably need to bite the bullet and migrate this code to a single configuration method for all install types.  This would be a one-time breaking event for manual Docker and standard install users, but I don't know how to avoid this really, although perhaps I can make some of the migration automatic.

My original plan was to do this as part of a 5.0 release that was also planned to be a re-write of this code into a more event based model but, as I no longer really have significant time to put into this project, and it mostly works well as is, I'd say a 5.0 release as originally planned is far enough away that I have to question if it will ever happen.

Perhaps the correct and simplest path for the short term is to just create a 5.0 release that does nothing more than implement these breaking configuration changes since that would lead to a much simplified code-path for configuration management as the rest of the changes are, while nice from a development perspective, don't really impact the user experiences in any way.

These are things I'll have to decide over the coming months based on the limited time I have available.

**Comment by tsightler** [comment]
BTW, there's another reason why the above code is not enough, it works for the NodeJS portion, but it doesn't work at all for the scripts that are used by the RTSP simple server to trigger stream events, etc.  These scripts leverage the mosquitto command line clients to connect to the MQTT broker and then use that as a control channel between the ring-mqtt NodeJS process and the rtsp-simple-server process.

So, to fully and properly support MQTTS and certificate based authentication the scripts would need to be modified to accept and pass different parameters to the mosquitto clients and to properly configure rtsp-simple-server with the required information (it's rtsp-simple-server itself that spawns the scripts on-demand when streaming clients connect).  Certainly the mosquitto clients support all of this, but my point is that it's a lot of work that would need to be taken care of and that is not addressed in this PR.

I'm going to put this on the roadmap for 5.1+, but, as 5.0 is still quite a ways out, it will unfortunately be a while.

**Comment by jamesonuk** [comment]
Just a a note for me for when I can look at this again but I just tested
`mosquitto_sub --cafile "${MQTTCA}" --cert "${MQTTCLIENTCERT}" --key "${MQTTCLIENTKEY}" -h "${MQTTHOST}" -p "${MQTTPORT}" -u "${MQTTUSER}" -P "${MQTTPASSWORD}" -t \zigbee2mqtt/#`
and that seemed to work fine (for both username / password and certificate).

I guess it is just the calls to mosquitto_pub and mosquitto_sub in https://github.com/tsightler/ring-mqtt/blob/main/scripts/start-stream.sh that need tweaking

**Comment by tsightler** [comment]
Thanks @jamesonuk.  I didn't think it would be too difficult, just a matter of finding the time for me to code it up and test it.  I'm getting there I think, slowly working on the 5.0 release which will standardize config/state management across all versions and, more importantly, hopefully allow for future device level states to be persistent across restarts, but TLS/certificate auth support is also on that list.

It needs to be added to the start-stream.sh script, and then the MQTT environment settings will need to be exported for the non-docker cases although exactly how to do that might change as part of the 5.0 changes (undecided so far).

**Pull request closed without merging**